engine/src/flutter/impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -50,496 +52,581 @@ class PositionWriter {
   }
 
-  Scalar direction = CreateBevelAndGetDirection(vtx_builder, position,
-                                                start_offset, end_offset);
+  inline void AppendVertices(const Point curve_point,
+                             SeparatedVector2 perpendicular) {
+    return AppendVertices(curve_point, perpendicular.GetVector());
+  }
 
-  Point miter_point = (((start_offset + end_offset) / 2) / alignment);
-  if (miter_point.GetDistanceSquared({0, 0}) > miter_limit * miter_limit) {
-    return;  // Convert to bevel when we exceed the miter limit.
+  inline void HandlePreviousJoin(SeparatedVector2 new_perpendicular) {
+    FML_DCHECK(has_prior_contour_);
+    if (has_prior_segment_) {
+      AddJoin(join_, last_point_, last_perpendicular_, new_perpendicular);
+    } else {
+      has_prior_segment_ = true;
+      auto perpendicular_vector = new_perpendicular.GetVector();
+      if (contour_needs_cap_) {
+        AddCap(cap_, last_point_, perpendicular_vector, true);
+      }
+      // Start the new segment's "box" at the shared "last_point_" with
+      // the new perpendicular vector.
+      AppendVertices(last_point_, perpendicular_vector);
+      origin_perpendicular_ = new_perpendicular;
+    }
   }
 
-  // Outer miter point.
-  vtx_builder.AppendVertex(position + miter_point * direction);
-}
+  // Adds a cap to an endpoint of a contour. The location points to the
+  // centerline of the stroke. The perpendicular points clockwise to the
+  // direction the path is traveling and is the length of half of the
+  // stroke width.
+  //
+  // If contour_start is true, then the cap is being added prior to the first
+  // segment at the beginning of a contour and assumes that no points have
+  // been added for this contour yet and also that the caller will add the
+  // two points that start the segment's "box" when this method returns.
+  //
+  // If contour_start is false, then the cap is being added after the last
+  // segment at the end of a contour and assumes that the caller has already
+  // added the two segments that define the end of the "box" for the last
+  // path segment.
+  void AddCap(Cap cap,
+              Point path_point,
+              Vector2 perpendicular,
+              bool contour_start) {
+    switch (cap) {
+      case Cap::kButt:
+        break;
+      case Cap::kRound: {
+        Point along(perpendicular.y, -perpendicular.x);
+        if (contour_start) {
+          // Start with a single point at the far end of the round cap
+          vtx_builder_.AppendVertex(path_point - along);
+
+          // Iterate down to, but not including, the first entry (1, 0)
+          for (size_t i = trigs_.size() - 2u; i > 0u; --i) {
+            Point center = path_point - along * trigs_[i].sin;
+            Vector2 offset = perpendicular * trigs_[i].cos;
+
+            AppendVertices(center, offset);
+          }
+        } else {
+          // Iterate up to, but not including, the last entry (0, 1)
```

Comment:
```suggestion
          // Iterate up to, but not including, the last entry (0, 1).
```

Reply:
It's not well documented, but Trigs always starts with (0, 1) and ends with (1, 0). The minimum divisions is 1 which means go all the way in one step which ends up with 2 entries (0,1) and (1, 0). I'm not sure where to best document all of this. Perhaps just FML_DCHECK in the constructor that size > 1 and put a comment on that?